scripts/machine_util.py

```diff
@@ -0,0 +1,120 @@
+"""
+import argparse
+import subprocess
+import sys
+from pathlib import Path
+
+def read_sys_file(sysfile: Path):
+    with open(sysfile, 'r') as f:
+        return f.read()
+
+def write_sys_file(sysfile: Path, content: str):
+    print(f"Write {content} to {sysfile}")
+    with open(sysfile, 'w') as f:
+        f.write(content)
+
+def check_intel_turbo_state(turbo_file='/sys/devices/system/cpu/intel_pstate/no_turbo'):
+    return int(read_sys_file(turbo_file))
+
+def set_intel_turbo_state(state: int, turbo_file='/sys/devices/system/cpu/intel_pstate/no_turbo'):
+    assert state in [0, 1]
+    write_sys_file(turbo_file, str(state))
+
+def parse_lscpu_cpu_core_list():
+    coreinfo = subprocess.check_output("lscpu --all --parse=CPU,CORE,ONLINE", shell=True).strip().decode().split('\n')
+    matched_cpus = 0
+    cpu_core = []
+    for line in coreinfo[2:]:
+        if line[0] == '#':
+            continue
+        cpu, core, online = line.split(',')
+        cpu = int(cpu)
+        online = online == "Y"
+        core = int(core) if online else None
+        if cpu == core:
+            matched_cpus += 1
+        cpu_core.append((cpu, core, online))
+    assert matched_cpus > 0, "Failed to parse lscpu output"
+    return cpu_core
+
+
+def hyper_threading_enabled():
+    for cpu, core, online in parse_lscpu_cpu_core_list():
+        if cpu != core and online:
+            return True
+    return False
+
+def set_hyper_threading(enabled=False):
+    for cpu, core, online in parse_lscpu_cpu_core_list():
+        if cpu != core:
+            if not online and not enabled:
+                continue
+            if online and enabled:
+                continue
+            virtual_cpu_online_file = f"/sys/devices/system/cpu/cpu{cpu}/online"
+            value = "1" if enabled else "0"
+            write_sys_file(virtual_cpu_online_file, value)
+
+def get_intel_max_cstate():
+    kernel_args = read_sys_file('/proc/cmdline').split()
+    for arg in kernel_args:
+        if arg.find('intel_idle.max_cstate') == 0:
+            return int(arg.split('=')[1])
+    return None
+
+def get_isolated_cpus():
+    """
+    Returns a list of cpus marked as isolated from the kernel scheduler for regular tasks.
+    Only tasks scheduled via taskset command can use these cpus, e.g. benchmarking workload.
+    """
+    kernel_args = read_sys_file('/proc/cmdline').split()
+    isolcpus = set()
+    for arg in kernel_args:
+        if arg.find('isolcpus') == 0:
+            arg = arg.split('=')[1]
+            chunks = arg.split(',')
+            for chunk in chunks:
+                if '-' in chunk:
+                    start, end = chunk.split('-')
+                    for cpu in range(start, end+1):
+                        isolcpus.add(cpu)
+                else:
+                    isolcpus.add(int(chunk))
+    return list(isolcpus)
+
+def get_nvidia_graphics_clock(device_id=0):
```

Comment:
Does it make sense to get all available graphics clock by default?

Reply:
yep, i have already made that change actually. I am adding more stuff still.  I'll update the PR in a little bit.